std/algorithm/comparison.d

```diff
@@ -892,116 +892,367 @@ nothrow pure @safe unittest
+
+    import std.typecons : Tuple;
+    Tuple!NewTypes ys = void;
+
+    alias fun_ = unaryFun!(fun);
+
+    import std.conv : emplace;
+    static foreach (immutable i, x; xs)
+    {
+        emplace(&ys[i], fun_(x));
+    }
+
+    return ys;
+}
+
 // equal
 /**
-Compares two ranges for equality, as defined by predicate `pred`
-(which is `==` by default).
+Compares two or more ranges `rs` for equality, as defined by predicate `pred`
+(which is `a == b` by default).
 */
 template equal(alias pred = "a == b")
 {
-    enum isEmptyRange(R) =
-        isInputRange!R && __traits(compiles, {static assert(R.empty, "");});
+    import std.meta : anySatisfy, allSatisfy;
+
+    /** Check if `T` is always empty (`empty` is an `enum` always being `true`)
+     */
+    static enum isAlwaysEmpty(T) = (__traits(compiles, { static assert(T.empty); }));
 
-    enum hasFixedLength(T) = hasLength!T || isNarrowString!T;
+    static enum areAllEquableRanges(Rs...) = is(typeof(binaryFun!pred(ElementType!(Rs[0]).init, // TODO check all `Rs`
+                                                                      ElementType!(Rs[1]).init)));
+
+    static alias ElementEncodingTypeUnqual(R) = Unqual!(ElementEncodingType!R);
+
+    static enum hasIndexingAndLength(T) = (is(typeof(T.init[0])) &&
+                                           is(typeof(T.length)));
+
+    /// TODO better name, anyone?
+    static auto maybeByCodeUnit(T)(T x)
+    {
+        static if (isAutodecodableString!T)
+        {
+            import std.utf : byCodeUnit;
+            return x.byCodeUnit;
+        }
+        else
+        {
+            return x;
+        }
+    }
 
     /++
-    Compares two ranges for equality. The ranges may have
-    different element types, as long as `pred(r1.front, r2.front)`
-    evaluates to `bool`.
-    Performs $(BIGOH min(r1.length, r2.length)) evaluations of `pred`.
+     This function compares two ore more ranges `rs` for equality. The
+    ranges may have different element types, as long as `pred(a, b)`
+    evaluates to `bool` for every combination elements of `a` and `b` at the same position in the ranges
+    `rs`.  Performs $(BIGOH min(r1.length, rs.length)) TODO BIG differently
+    evaluations of `pred`.
 
     Params:
-        r1 = The first range to be compared.
-        r2 = The second range to be compared.
+        rs = The ranges to be compared.
 
     Returns:
-        `true` if and only if the two ranges compare _equal element
+        `true` if and only if all the ranges compare _equal element
         for element, according to binary predicate `pred`.
     +/
-    bool equal(Range1, Range2)(Range1 r1, Range2 r2)
-    if (isInputRange!Range1 && isInputRange!Range2 &&
-        is(typeof(binaryFun!pred(r1.front, r2.front))))
+    bool equal(Rs...)(Rs rs)
+        if (rs.length >= 2 &&
+            areAllEquableRanges!(Rs) &&
+            allSatisfy!(isInputRange, Rs) &&
+            !allSatisfy!(isInfinite, Rs)
```

Comment:
Shouldn't this be `!anySatisfy(isInfinite, Rs)`. I mean as long as there's one infinite range in the mix, it's more or less ill-defined. this would save you a lot of special casing too.